packages/expo-doctor/src/utils/explainDependencies.ts

```diff
@@ -0,0 +1,132 @@
+async function explainAsync(
+  packageName: string,
+  projectRoot: string,
+  parameters: string[] = []
+): Promise<RootNodePackage[] | null> {
+  const args = ['explain', packageName, ...parameters, '--json'];
+
+  try {
+    const { stdout } = await spawnAsync('npm', args, {
+      stdio: 'pipe',
+      cwd: projectRoot,
+    });
+
+    return JSON.parse(stdout);
+  } catch (error: any) {
+    if (isSpawnResult(error)) {
+      if (error.stderr.match(/npm ERR! No dependencies found matching/)) {
+        return null;
+      } else if (error.stdout.match(/Usage: npm <command>/)) {
+        throw new Error(
+          `Dependency tree validation for ${chalk.underline(
+            packageName
+          )} failed. This validation is only available on Node 16+ / npm 8.`
+        );
+      }
+    }
+    throw new Error(`Failed to find dependency tree for ${packageName}: ` + error.message);
+  }
+}
+
+function organizeExplanations(
+  pkg: TargetPackage,
+  {
+    explanations,
+    isValid,
+  }: {
+    explanations: RootNodePackage[];
+    isValid: (pkg: TargetPackage) => boolean;
+  }
+) {
+  const valid: RootNodePackage[] = [];
+  const invalid: RootNodePackage[] = [];
+
+  for (const explanation of explanations) {
+    const { name } = explanation;
+    if (name === pkg.name) {
+      if (isValid(explanation)) {
+        valid.push(explanation);
+      } else {
+        invalid.push(explanation);
+      }
+    }
+  }
+  return { valid, invalid };
+}
+
+/**
+ * @param pkg
+ * @param explanations
+ * @returns true if all versions of the package satisfy the constraints
+ */
+async function printExplanationsAsync(
+  pkg: TargetPackage,
+  explanations: RootNodePackage[]
+): Promise<string | null> {
+  const { invalid } = organizeExplanations(pkg, {
+    explanations,
+    isValid(otherPkg) {
+      return semver.satisfies(otherPkg.version!, pkg.version!);
+    },
+  });
+
+  if (invalid.length > 0) {
+    return getInvalidPackagesWarning(pkg, { explanations: invalid });
+  }
+  return null;
+}
+
+function getInvalidPackagesWarning(
+  pkg: TargetPackage,
+  { explanations }: { explanations: RootNodePackage[] }
+): string {
+  const lines = [];
+  if (pkg.version) {
+    lines.push(`Expected package ${formatPkg(pkg, 'green')}`);
+  } else {
+    lines.push(`Expected to not find any copies of ${formatPkg(pkg, 'green')}`);
+  }
+  lines.push(chalk`Found invalid:`);
+  lines.push(explanations.map(explanation => '  ' + formatPkg(explanation, 'red')).join('\n'));
+  lines.push(chalk`  {dim (for more info, run: {bold npm why ${pkg.name}})}`);
+
+  return lines.join('\n');
+}
+
+function formatPkg(pkg: TargetPackage, versionColor: string) {
+  if (pkg.version) {
+    return chalk`{bold ${pkg.name}}{cyan @}{${versionColor} ${pkg.version}}`;
+  } else {
+    return chalk`{bold ${pkg.name}}`;
+  }
+}
+
+/**
+ * @param pkg
+ * @returns true if all versions of the package satisfy the constraints
```

Comment:
incorrect return value doc. Does not return true, returns `null` if valid, otherwise a string.